File: apps/ios/ExpCore/Sources/Domain/AttachmentFiles.swift
import Foundation

/// EXP-297 — the client-side half of the attachment classification contract.
///
/// Attachments are no longer image-only: any content type can be uploaded (10 MB
/// for the inline-image types, 50 MB for everything else). A row is an INLINE
/// IMAGE iff its `content_type` is one of the five raster types the markdown
/// pipeline accepts — those live in the description/comment markdown as
/// `![](…)` and are rendered by the editor. EVERY other row (including other
/// `image/*` types such as tiff) belongs in the issue's Files section, which
/// renders straight from the synced `attachments` rows.
///
/// Mirrors `acceptedImageContentTypes` / `maxFileUploadBytes` in
/// apps/web/src/lib/storage/issue-attachments.ts, and its Android twin
/// `domain/AttachmentFiles.kt`.
public enum AttachmentFiles {
    /// The five content types the inline-image pipeline accepts. Exact mirror of
    /// the server's `acceptedImageContentTypes` — do not widen without changing
    /// the server first, or files would silently vanish from the Files section.
    public static let inlineImageContentTypes: Set<String> = [
        "image/png",
        "image/jpeg",
        "image/webp",
        "image/gif",
        "image/avif",
    ]

    /// 50 MB — the server's cap for non-image uploads (`maxFileUploadBytes`).
    /// Checked client-side so an oversize pick fails instantly instead of after
    /// a long upload.
    public static let maxFileUploadBytes = 50 * 1024 * 1024

    /// Fallback content type for a pick whose extension maps to no UTI.
    public static let fallbackContentType = "application/octet-stream"

    /// EXACT string match, mirroring the server, web, and desktop — a stored
    /// `image/PNG` or parameterized type is a Files row everywhere, and
    /// classifying it inline here would hide it on iOS only. Client-derived
    /// picker types must be canonicalized via `canonicalContentType` BEFORE
    /// upload, never at classification time.
    public static func isInlineImage(contentType: String) -> Bool {
        inlineImageContentTypes.contains(contentType)
    }

    /// Canonical upload form of a picker-derived content type: lowercased media
    /// essence with any `;`-parameter suffix stripped, falling back to
    /// `application/octet-stream`. Mirrors the server's
    /// `canonicalizeContentType` so stored rows classify identically on every
    /// client.
    public static func canonicalContentType(_ raw: String?) -> String {
        let essence = normalized(raw ?? "")
        return essence.isEmpty ? fallbackContentType : essence
    }

    /// SF Symbol for a file row's leading type glyph. Deliberately NOT a Lucide
    /// registry icon: the shared registry carries no file-type glyphs, and these
    /// are decorative per-row hints rather than a cross-client unified surface.
    public static func sfSymbolName(forContentType contentType: String) -> String {
        let type = normalized(contentType)
        if type == "application/pdf" { return "doc.richtext" }
        if isArchive(type) { return "doc.zipper" }
        if type.hasPrefix("video/") { return "film" }
        if type.hasPrefix("audio/") { return "waveform" }
        if type.hasPrefix("text/") { return "doc.text" }
        return "doc"
    }

    /// A filename safe to write into a temp directory: path separators and
    /// traversal segments can't escape the per-attachment folder, and an empty
    /// result falls back to a generic name.
    public static func sanitizedFilename(_ filename: String) -> String {
        let cleaned = String(filename.map { character in
            let scalar = character.unicodeScalars.first
            let isControl = scalar.map { CharacterSet.controlCharacters.contains($0) } ?? false
            return (character == "/" || character == ":" || character == "\\" || isControl)
                ? "_"
                : character
        })
        let trimmed = cleaned.trimmingCharacters(in: .whitespacesAndNewlines)
        if trimmed.isEmpty || trimmed == "." || trimmed == ".." { return "file" }
        // Keep the name comfortably inside every filesystem's per-component cap.
        return String(trimmed.prefix(120))
    }

    // MARK: - Internals

    /// Lowercased, parameter-free (`text/plain; charset=utf-8` → `text/plain`).
    private static func normalized(_ contentType: String) -> String {
        let base = contentType.split(separator: ";", maxSplits: 1).first.map(String.init) ?? contentType
        return base.trimmingCharacters(in: .whitespaces).lowercased()
    }

    private static let archiveContentTypes: Set<String> = [
        "application/zip",
        "application/x-zip-compressed",
        "application/gzip",
        "application/x-gzip",
        "application/x-tar",
        "application/x-bzip2",
        "application/x-7z-compressed",
        "application/vnd.rar",
        "application/x-rar-compressed",
    ]

    private static func isArchive(_ type: String) -> Bool {
        archiveContentTypes.contains(type)
    }
}

File: apps/android/app/src/main/java/com/exponential/app/domain/AttachmentFiles.kt
package com.exponential.app.domain

/**
 * Attachment classification + file-handling helpers (EXP-297).
 *
 * An attachment row is an INLINE IMAGE iff its `content_type` is exactly one
 * of the five raster types the markdown pipeline embeds — the exact mirror of
 * the server's `acceptedImageContentTypes` (apps/web/src/lib/storage/issue-attachments.ts).
 * Everything else — including any other `image/` subtype such as `image/tiff` or
 * `image/svg+xml` — belongs in the issue's Files section, so nothing an upload
 * accepted can end up invisible.
 */
val INLINE_IMAGE_CONTENT_TYPES: Set<String> = setOf(
    "image/png",
    "image/jpeg",
    "image/webp",
    "image/gif",
    "image/avif",
)

/**
 * True when [contentType] names one of the five inline-embeddable raster
 * types. The comparison is an EXACT string match, mirroring the server, web,
 * and desktop — a stored `image/PNG` or parameterized type is a Files row
 * everywhere, and classifying it inline here would hide it on Android only.
 * Client-derived types (`ContentResolver.getType`) must be canonicalized via
 * [canonicalContentType] BEFORE upload, never at classification time.
 */
fun isInlineImage(contentType: String?): Boolean =
    contentType != null && contentType in INLINE_IMAGE_CONTENT_TYPES

/**
 * Canonical upload form of a picker-derived content type: lowercased media
 * essence with any `;`-parameter suffix stripped, falling back to
 * `application/octet-stream`. Mirrors the server's `canonicalizeContentType`
 * so stored rows always classify identically on every client.
 */
fun canonicalContentType(raw: String?): String {
    val essence = (raw ?: "").substringBefore(';').trim().lowercase()
    return essence.ifEmpty { "application/octet-stream" }
}

/** Non-image upload cap (the server's `maxFileUploadBytes`). */
const val MAX_FILE_UPLOAD_BYTES: Long = 50L * 1024 * 1024

/**
 * Inline-image upload cap (the server's `maxImageUploadBytes`) — refuse
 * locally rather than push megabytes over a mobile uplink only to be rejected.
 * Shared by the steer composer (EXP-511) and comment attachments (EXP-554).
 */
const val MAX_IMAGE_UPLOAD_BYTES: Long = 10L * 1024 * 1024

/** Attachments one comment can carry (the server's `MAX_COMMENT_ATTACHMENTS`). */
const val MAX_COMMENT_ATTACHMENTS: Int = 10

/**
 * Make a server-supplied filename safe to use as a local cache filename:
 * strip path separators (so `../../x` can never escape the per-attachment
 * directory), control characters, and the path-traversal names themselves.
 * Clamped so long names can't blow the filesystem's per-component limit.
 */
fun sanitizeFilename(name: String?): String {
    val cleaned = (name ?: "")
        .map { ch ->
            when {
                ch == '/' || ch == '\\' -> '_'
                ch.code < 0x20 || ch.code == 0x7F -> '_'
                else -> ch
            }
        }
        .joinToString("")
        .trim()
    if (cleaned.isEmpty() || cleaned == "." || cleaned == "..") return "file"
    return if (cleaned.length > MAX_FILENAME_LENGTH) cleaned.take(MAX_FILENAME_LENGTH) else cleaned
}

private const val MAX_FILENAME_LENGTH = 120
